src/VkLayer_GF_amber_scoop/include/VkLayer_GF_amber_scoop/amber_scoop_layer.h

```diff
@@ -62,6 +66,9 @@ struct DeviceData {
   // Tracked device data:
 
   ProtectedMap<VkCommandBuffer, CommandBufferData> command_buffers_data;
+  ProtectedMap<VkPipeline, VkGraphicsPipelineCreateInfo> graphics_pipelines;
+  ProtectedMap<VkShaderModule, std::unique_ptr<ShaderModuleData>>
+      shader_modules_data;
```

Comment:
I think we should have a `GraphicsPipelineData` class, which will take a `const VkGraphicsPipelineCreateInfo&` in the constructor, which will be deep copied, and the destructor will `DeepDelete` it.

I think `ShaderModuleData` should similarly take a `const VkShaderModuleCreateInfo&` in its constructor and `DeepCopy` it.

Basically, I think all (most?) calls to `DeepCopy` should be in a constructor, with `DeepDelete` in the corresponding destructor.

Reply:
That makes sense. I think there will be great amount of those classes with only create info as a field and constructor and destructor implemented (and getter for the create info), so should we add all those in a single .h file or one per class? Or maybe even use a template class like:
```c++
template <typename CreateInfoType>
class CreateInfoWrapper {
 public:
  explicit CreateInfoWrapper(const CreateInfoType &create_info) : create_info_(DeepCopy(create_info)) {}

  ~CreateInfoWrapper() {
    DeepDelete(create_info_);
  }

  CreateInfoType* GetCreateInfo() {
    return &create_info_;
  }

 private:
  CreateInfoType create_info_;
};
```

Reply:
I don't mind. Up to you.